File: backend/application/api/exceptions.py
"""API Runtime Exceptions (Phase 15.1).

Defines the exception hierarchy for API runtime operations, provider errors,
initialization failures, configuration issues, and validation errors.
"""


class ApiRuntimeException(Exception):
    """Base exception for all API runtime errors."""

    pass


class ApiInitializationException(ApiRuntimeException):
    """Raised when API runtime or provider initialization fails."""

    pass


class ApiConfigurationException(ApiRuntimeException):
    """Raised when API configuration is invalid or missing."""

    pass


class ApiProviderException(ApiRuntimeException):
    """Raised when API provider operations encounter an error."""

    pass


class ApiValidationException(ApiRuntimeException):
    """Raised when API runtime validation checks fail."""

    pass

File: frontend/src/runtime/events/exceptions.ts
/**
 * Event Runtime Exception Hierarchy (Phase 16.4.1).
 *
 * Defines custom exceptions for event runtime initialization, provider errors,
 * dispatch failures, and event validation errors.
 */

export class EventRuntimeException extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'EventRuntimeException';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class EventInitializationException extends EventRuntimeException {
  constructor(message: string) {
    super(message);
    this.name = 'EventInitializationException';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class EventProviderException extends EventRuntimeException {
  constructor(message: string) {
    super(message);
    this.name = 'EventProviderException';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class EventDispatchException extends EventRuntimeException {
  constructor(message: string) {
    super(message);
    this.name = 'EventDispatchException';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class EventValidationException extends EventRuntimeException {
  constructor(message: string) {
    super(message);
    this.name = 'EventValidationException';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
